src/tactic/induction.lean

```diff
@@ -0,0 +1,1290 @@
+/-
+Copyright (c) 2020 Jannis Limperg. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Jannis Limperg
+-/
+
+import tactic.dependencies
+import tactic.fresh_names
+import tactic.generalizes
+import tactic.has_variable_names
+import tactic.unify_equations
+
+/-!
+# A better tactic for induction and case analysis
+
+This module defines the tactics `tactic.interactive.induction'` and
+`tactic.interactive.cases'`, which are variations on Lean's builtin `induction`
+and `cases`. The primed variants feature various improvements over the builtin
+tactics; in particular, they generate more human-friendly names and `induction'`
+deals much better with indexed inductive types. See the tactics' documentation
+for more details. We also provide corresponding non-interactive induction
+tactics `tactic.eliminate_hyp` and `tactic.eliminate_expr`.
+
+The design and implementation of these tactics is described in a
+[draft paper](https://limperg.de/paper/cpp2021-induction/).
+-/
+
+open expr native
+open tactic.interactive (case_tag.from_tag_hyps)
+
+namespace tactic
+namespace eliminate
+
+
+/-!
+## Tracing
+
+We set up two tracing functions to be used by `eliminate_hyp` and its supporting
+tactics. Their output is enabled by setting `trace.eliminate_hyp` to `true`.
+-/
+
+
+declare_trace eliminate_hyp
+
+/--
+`trace_eliminate_hyp msg` traces `msg` if the option `trace.eliminate_hyp` is
+`true`.
+-/
+meta def trace_eliminate_hyp {α} [has_to_format α] (msg : thunk α) : tactic unit :=
+when_tracing `eliminate_hyp $ trace $ to_fmt "eliminate_hyp: " ++ to_fmt (msg ())
+
+/--
+`trace_state_eliminate_hyp msg` traces `msg` followed by the tactic state if the
+option `trace.eliminate_hyp` is `true`.
+-/
+meta def trace_state_eliminate_hyp {α} [has_to_format α] (msg : thunk α) :
+  tactic unit := do
+  state ← read,
+  trace_eliminate_hyp $ format.join
+    [to_fmt (msg ()), "\n-----\n", to_fmt state, "\n-----"]
+
+
+/-!
+## Information Gathering
+
+We define data structures for information relevant to the induction, and
+functions to collect this information for a specific goal.
+-/
+
+
+/--
+Information about a constructor argument. Contains:
```

Comment:
Explain what you mean by a constructor argument.

Reply:
Will do.